File: javascript/hamming/hamming.js
module.exports = class Hamming {
  compute(firstStrand, secondStrand) {
    if (firstStrand.length !== secondStrand.length)
      throw new Error("DNA strands must be of equal length.");
    this._makeSequences(firstStrand, secondStrand);
    return this._countDifferences();
  }

  _makeSequences(firstStrand, secondStrand) {
    this._expandedFirstStrand = firstStrand.split("");
    this._expandedSecondStrand = secondStrand.split("");
  }

  _countDifferences() {
    return this._expandedFirstStrand.reduce(
      (numberOfDifferences, letter, index) =>
        letter !== this._expandedSecondStrand[index]
          ? numberOfDifferences + 1
          : numberOfDifferences,
      0
    );
  }
};

File: java/hamming/src/main/java/Hamming.java
class Hamming {
    private String leftStrand;
    private String rightStrand;
    private int distance;

    Hamming(String leftStrand, String rightStrand) {
        if (leftStrand.length() != rightStrand.length())
            throw new IllegalArgumentException("leftStrand and rightStrand must be of equal length.");

        this.leftStrand = leftStrand;
        this.rightStrand = rightStrand;
        this.distance = calculateHammingDistance();
    }

    int getHammingDistance() {
        return distance;
    }

    private int calculateHammingDistance() {
        for (int i = 0, length = leftStrand.length(); i < length; i++)
            if (isNotSameCharacter(i))
                distance++;
        return distance;
    }

    private boolean isNotSameCharacter(int index) {
        return leftStrand.charAt(index) != rightStrand.charAt(index);
    }
}
